include/swift/AST/DiagnosticsSema.def

```diff
@@ -3852,6 +3852,9 @@ ERROR(unresolved_member_no_inference,none,
       (DeclNameRef))
 ERROR(cannot_infer_base_of_unresolved_member,none,
       "cannot infer contextual base in reference to member %0", (DeclNameRef))
+ERROR(cannot_infer_concrete_for_opaque_result,none,
+      "cannot infer concrete type for opaque result %0 from return expression",
+      (Type))
```

Comment:
Some thoughts:

* We use "underlying type" rather than "concrete type" in the [book](https://docs.swift.org/swift-book/LanguageGuide/OpaqueTypes.html), and in [other diagnostics](https://github.com/apple/swift/blob/eecde02dde62424bc8a93de09f30741cf6f946e8/include/swift/AST/DiagnosticsSema.def#L4415-L4439) about opaque result types. We ought to be consistent.
* Should we align the wording of this message with the more general `generic parameter '%0' could not be inferred`? For example, `underlying type for opaque result type %0 could not be inferred from return expression`. They are the same diagnostic at the core, after all. I think the associations that stem from having these messages in sync would support our efforts in presenting this feature as an extension of the generics system.

Reply:
> We use "underlying type" rather than "concrete type" in the [book](https://docs.swift.org/swift-book/LanguageGuide/OpaqueTypes.html), and in [other diagnostics](https://github.com/apple/swift/blob/eecde02dde62424bc8a93de09f30741cf6f946e8/include/swift/AST/DiagnosticsSema.def#L4415-L4439) about opaque result types. We ought to be consistent.

Makes sense :)

> * Should we align the wording of this message with the more general `generic parameter '%0' could not be inferred`?

This was based in other places like `cannot infer return type of empty closure` and if we search for "cannot infer" in all *.def files we see a few others where this wording is used. So I wonder if makes sense to audit them all to see where such change makes sense as well? 

> For example, `underlying type for opaque result type %0 could not be inferred from return expression`. They are the same diagnostic at the core, after all. I think the associations that stem from having these messages in sync would support our efforts in presenting this feature as an extension of the generics system.

Sounds reasonable to me! I guess this would give a good first issue

Reply:
Good first issue: #63291

Reply:
Awesome, thanks!